contracts/RelayHub.sol

```diff
@@ -213,17 +213,17 @@ contract RelayHub is RelayHubApi {
         // gas_reserve must be high enough to complete relay()'s post-call execution.
         require(safe_sub(initial_gas,gas_limit) >= gas_reserve, "Not enough gasleft()");
         bool success_post;
-        bytes memory ret;
+        bytes memory ret = new bytes(32);
         (success_post,ret) = address(this).call(abi.encodeWithSelector(this.recipient_calls.selector,from,to,msg.sender,encoded_function,transaction_fee,gas_limit,initial_gas));
         nonces[from]++;
+        RelayCallStatus status = RelayCallStatus.OK;
+        if (LibBytes.readUint256(ret,0) == 0)
+            status = RelayCallStatus.RelayedCallFailed;
         // Relay transaction_fee is in %.  E.g. if transaction_fee=40, payment will be 1.4*used_gas.
         uint charge = (gas_overhead+initial_gas-gasleft())*gas_price*(100+transaction_fee)/100;
         if (!success_post){
             emit TransactionRelayed(msg.sender, from, to, keccak256(encoded_function), uint(RelayCallStatus.PostRelayedFailed), charge);
```

Comment:
You still have "if", and still do dynamic-bytes method (`keccak`) AFTER the calculation.

the keccak (and retStatus) should be filled before the charge calculation, and thus the emit looks like:

```
 emit TransactionRelayed(msg.sender, from, to, hash_func, retStatus, charge);
```
(single emit, with no `if`)

actually, the keccak should be completely REMOVED: the method hash is meaningless:  we never use it, and it can only detect the very rare an uninsteresting case of repeated function calls with same parameters...